c_binding/lsm_datatypes.cpp

```diff
@@ -2208,6 +2209,96 @@ char *wwpn_convert(const char *wwpn)
+                                      const char *system_id,
+                                      const char *plugin_data)
+{
+    lsm_battery *rc = NULL;
+
+    if ((id == NULL) || (name == NULL) || (system_id == NULL))
+        return NULL;
+
+    rc = (lsm_battery *) malloc(sizeof(lsm_battery));
+    if (rc != NULL) {
+        rc->magic = LSM_BATTERY_MAGIC;
+        rc->id = strdup(id);
+        rc->name = strdup(name);
+        rc->type = type;
+        rc->status = status;
+        rc->system_id = strdup(system_id);
+        rc->plugin_data = NULL;
+
+        if (plugin_data != NULL) {
+            rc->plugin_data = strdup(plugin_data);
+            if (rc->plugin_data == NULL) {
+                lsm_battery_record_free(rc);
+                rc = NULL;
+            }
+        }
+
+        if (rc->id == NULL || rc->name == NULL || rc->system_id == NULL) {
+            lsm_battery_record_free(rc);
+            rc = NULL;
+        }
+    }
+    return rc;
+}
+
+int lsm_battery_record_free(lsm_battery *b)
+{
+    if (LSM_IS_BATTERY(b)) {
+        b->magic = LSM_DEL_MAGIC(LSM_BATTERY_MAGIC);
+        free(b->name);
+        b->name = NULL;
+        free(b->id);
+        b->id = NULL;
+        free(b->system_id);
+        b->system_id = NULL;
+        free(b->plugin_data);
+        b->plugin_data = NULL;
+        free(b);
+        return LSM_ERR_OK;
+    }
+    return LSM_ERR_INVALID_ARGUMENT;
+}
+
+lsm_battery *lsm_battery_record_copy(lsm_battery *b)
+{
+    if (LSM_IS_BATTERY(b))
+        return lsm_battery_record_alloc(b->id, b->name, b->type, b->status,
+                                        b->system_id, b->plugin_data);
+    return NULL;
+}
+
+
+CREATE_FREE_ARRAY_FUNC(lsm_battery_record_array_free, lsm_battery_record_free,
+                       lsm_battery *, LSM_ERR_INVALID_ARGUMENT);
+
+#define CREATE_PRO_GETTER_FUNC(func_name, obj_type, struct_check_func, \
+                               pro_type, pro_name, fail_value) \
+pro_type func_name(obj_type obj) \
+{ \
+    MEMBER_GET(obj, struct_check_func, pro_name, fail_value); \
+}
+
+CREATE_PRO_GETTER_FUNC(lsm_battery_id_get, lsm_battery *, LSM_IS_BATTERY,
+                       const char *, id, NULL);
+CREATE_PRO_GETTER_FUNC(lsm_battery_name_get, lsm_battery *, LSM_IS_BATTERY,
+                       const char *, name, NULL);
+CREATE_PRO_GETTER_FUNC(lsm_battery_system_id_get, lsm_battery *, LSM_IS_BATTERY,
+                       const char *, system_id, NULL);
+CREATE_PRO_GETTER_FUNC(lsm_battery_plugin_data_get, lsm_battery *,
+                       LSM_IS_BATTERY, const char *, plugin_data, NULL);
+CREATE_PRO_GETTER_FUNC(lsm_battery_status_get, lsm_battery *, LSM_IS_BATTERY,
+                       uint64_t, status, LSM_BATTERY_STATUS_UNKNOWN);
+CREATE_PRO_GETTER_FUNC(lsm_battery_type_get, lsm_battery *,
+                       LSM_IS_BATTERY, lsm_battery_type, type,
+                       LSM_BATTERY_TYPE_UNKNOWN);
+
```

Comment:
We have a `#define` already `MEMBER_FUNC_GET`, can we re-use that one instead of creating the new one `CREATE_PRO_GETTER_FUNC`?